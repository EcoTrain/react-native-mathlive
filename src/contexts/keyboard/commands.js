import {useContext} from 'react';
import {MathfieldContext} from '../mathfield/MathfieldContext';

const COMMANDS = {};

/**
 * Execute command on key press
 */
export const executeCommand = ({command, options}) => {
  const _command = COMMANDS[command];
  if (_command) {
    _command({options});
  }
};

/**
 * Define commands for executeCommand on key press
 */
export const defineCommands = () => {
  const mfc = useContext(MathfieldContext);

  const getPreviousAtom = () => {
    let prevAtom;
    if (mfc.selectedAtom?.parent) {
      const leftSibling = mfc.selectedAtom.leftSibling;
      if (leftSibling) {
        prevAtom = leftSibling;
      } else {
        prevAtom = mfc.selectedAtom.parent;
      }
    } else {
      const ind = mfc.atoms.indexOf(mfc.selectedAtom);
      if (ind > 0) {
        prevAtom = mfc.atoms[ind - 1];
      }
    }
    return prevAtom;
  };
  const getNextAtom = () => {
    let nextAtom;
    if (mfc.selectedAtom?.parent) {
      const rightSibling = mfc.selectedAtom.rightSibling;
      if (rightSibling) {
        nextAtom = rightSibling;
      } else {
        nextAtom = mfc.selectedAtom.parent;
      }
    } else {
      const ind = mfc.atoms.indexOf(mfc.selectedAtom);
      if (ind > -1 && ind < mfc.atoms.length - 1) {
        nextAtom = mfc.atoms[ind + 1];
      }
    }
    return nextAtom;
  };

  COMMANDS.deleteBackward = () => {
    // TODO: change latex value on remove atom
    // TODO: add serialize to atoms
    if (mfc.selectedAtom?.parent) {
      console.log('deleteBackward atom child', {selectedAtom: mfc.selectedAtom, parent: mfc.selectedAtom.parent});
      const leftSibling = mfc.selectedAtom.leftSibling;

      // Set previous atom as selected
      if (leftSibling) {
        mfc.setSelectedAtom(leftSibling);
      } else {
        mfc.setSelectedAtom(mfc.selectedAtom.parent);
      }

      // Remove selected atom
      mfc.selectedAtom.parent.removeChild(mfc.selectedAtom);
    } else {
      const ind = mfc.atoms.indexOf(mfc.selectedAtom);
      console.log('deleteBackward atom root', {atoms: mfc.atoms, selectedAtom: mfc.selectedAtom, ind});
      if (ind > -1) {
        mfc.atoms.splice(ind, 1);
        // Set prev atom: Undefined if rm first
        mfc.setSelectedAtom(mfc.atoms[ind - 1]);
      }
    }
  };
  COMMANDS.addAtoms = ({options}) => {
    const _atoms = options.atoms || [];
    console.log('COMMANDS addAtoms', {atoms: mfc.atoms, _atoms, selectedAtom: mfc.selectedAtom});

    if (mfc.selectedAtom?.parent) {
      mfc.selectedAtom.parent.addChildrenAfter(_atoms, mfc.selectedAtom);
      if (mfc.selectedAtom.type == 'placeholder') {
        mfc.selectedAtom.parent.removeChild(mfc.selectedAtom);
      }
    } else {
      let ind = mfc.atoms.indexOf(mfc.selectedAtom);
      if (mfc.selectedAtom.type == 'placeholder') {
        mfc.atoms.splice(ind, 1);
        --ind;
      }
      mfc.atoms.splice(ind + 1, 0, ..._atoms);
    }
    mfc.setSelectedAtom(_atoms[_atoms.length - 1]);
  };
  COMMANDS.moveToPreviousChar = () => {
    const prevAtom = getPreviousAtom();
    prevAtom && mfc.setSelectedAtom(prevAtom);
  };
  COMMANDS.moveToNextChar = () => {
    const nextAtom = getNextAtom();
    nextAtom && mfc.setSelectedAtom(nextAtom);
  };
};
