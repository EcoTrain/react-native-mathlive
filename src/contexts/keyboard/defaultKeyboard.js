export const DEFAULT_KEYBOARD_LAYERS = {
  'geometry-layer': {
    rows: [
      [
        {latex: '\\vert'},
        {latex: '\\not{\\vert}'},
        {latex: '\\parallel'},
        {latex: '\\bot'},
        {latex: '\\overrightarrow{#@}'},
        {latex: '\\begin{pmatrix}#0\\\\#0\\end{pmatrix}'},
      ],
      [
        {latex: '\\measuredangle'},
        {latex: ';'},
        {latex: '\\longmapsto'},
        {latex: '\\left\\lvert #0 \\right\\rvert'},
        {latex: '\\left\\lvert \\overline{#0} \\right\\rvert'},
        {latex: '\\overline{#@}'},
      ],
      [
        {type: 'separator w20'},
        {type: 'separator w20'},
        {type: 'separator w20'},
        {type: 'separator w20'},
        {
          type: 'action',
          label: 'ᐸ',
          command: 'moveToPreviousChar',
        },
        {
          type: 'action',
          label: 'ᐳ',
          command: 'moveToNextChar',
        },
      ],
    ],
  },
  'basic-layer': {
    rows: [
      [
        {label: '7', insert: '7'},
        {label: '8', insert: '8'},
        {label: '9', insert: '9'},
        {label: '+', insert: '+'},
        {type: 'separator w5'},
        {label: '(', insert: '('},
        {label: ')', insert: ')'},
      ],
      [
        {label: '4', insert: '4'},
        {label: '5', insert: '5'},
        {label: '6', insert: '6'},
        {label: '-', insert: '-'},
        {type: 'separator w10'},
        {type: 'separator w5'},
        {latex: '\\sqrt{#?}'},
      ],
      [
        {label: '1', insert: '1'},
        {label: '2', insert: '2'},
        {label: '3', insert: '3'},
        {label: '*', insert: '*'},
        {type: 'separator w5'},
        {type: 'separator w10'},
        {
          type: 'action',
          label: 'x',
          command: 'deleteBackward',
        },
      ],
      [
        {type: 'separator w10'},
        {label: '0', insert: '0'},
        {label: ',', insert: ','},
        // TODO: Fix parse \ and as latex :
        {type: 'small', latex: '\\frac{#?}{#?}'},
        {type: 'separator w5'},
        {
          type: 'action',
          label: 'ᐸ',
          command: 'moveToPreviousChar',
        },
        {
          type: 'action',
          label: 'ᐳ',
          command: 'moveToNextChar',
        },
      ],
    ],
  },
};

export const DEFAULT_KEYBOARDS = {
  basic: {
    label: 'Basic',
    layer: 'basic-layer',
  },
  geometry: {
    label: 'Geometry',
    layer: 'geometry-layer',
  },
};
