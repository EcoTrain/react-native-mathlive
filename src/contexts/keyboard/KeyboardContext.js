import React, {createContext, useContext, useEffect, useState} from 'react';
import {MathfieldContext} from '../mathfield/MathfieldContext';
import {defineCommands} from './commands';
import {DEFAULT_KEYBOARDS, DEFAULT_KEYBOARD_LAYERS} from './defaultKeyboard';

export const KeyboardContext = createContext({
  isVisible: false,
  showKeyboard: () => {},
  hideKeyboard: () => {},
  toggleKeyboardVisibility: () => {},
  virtualKeyboardLayers: DEFAULT_KEYBOARD_LAYERS || {},
  virtualKeyboards: DEFAULT_KEYBOARDS || {},
  actualKeyboards: Object.keys(DEFAULT_KEYBOARDS) || {},
  activeKeyboardName: Object.keys(DEFAULT_KEYBOARDS)[0] || '',
  setActiveKeyboardName: () => {},
});

export const KeyboardContextProvider = ({children, kbConfig}) => {
  const [isVisible, setIsVisible] = useState(false);
  const {setFocus} = useContext(MathfieldContext);

  /**
   * Define commands for executeCommand on key press
   * (Redefine on every update context)
   */
  defineCommands();

  const virtualKeyboardLayers = kbConfig.mergeKeyboards
    ? Object.assign({}, DEFAULT_KEYBOARD_LAYERS, kbConfig.customKeyboardLayers)
    : DEFAULT_KEYBOARD_LAYERS;
  const virtualKeyboards = kbConfig.mergeKeyboards
    ? Object.assign({}, DEFAULT_KEYBOARDS, kbConfig.customKeyboards)
    : DEFAULT_KEYBOARDS;
  const actualKeyboards = Object.keys(virtualKeyboards);
  const [activeKbName, setActiveKbName] = useState(actualKeyboards[0]);

  const defaultContextValues = {
    isVisible,
    showKeyboard: () => {
      setIsVisible(true);
      setFocus(true);
    },
    hideKeyboard: () => {
      setIsVisible(false);
      setFocus(false);
    },
    toggleKeyboardVisibility: () => {
      setIsVisible(!isVisible);
      setFocus(!isVisible);
    },
    virtualKeyboardLayers,
    virtualKeyboards,
    actualKeyboards: actualKeyboards,
    activeKeyboardName: activeKbName,
    setActiveKeyboardName: setActiveKbName,
  };

  return <KeyboardContext.Provider value={defaultContextValues}>{children}</KeyboardContext.Provider>;
};
