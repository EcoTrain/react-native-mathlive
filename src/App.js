import React from 'react';
import { KeyboardContextProvider } from './contexts/keyboard/KeyboardContext';
import { MathfieldContextProvider } from './contexts/MathfieldContext';
import { ThemeContextProvider } from './contexts/ThemeContext';
import { VirtualKeyboard } from './editor/keyboard';
import { Mathfield } from './editor/mathfield';

export const App = () => {
  return (
    <ThemeContextProvider>
      <MathfieldContextProvider>
        <KeyboardContextProvider>
          <Mathfield />
          <VirtualKeyboard />
        </KeyboardContextProvider>
      </MathfieldContextProvider>
    </ThemeContextProvider>
  );
};
