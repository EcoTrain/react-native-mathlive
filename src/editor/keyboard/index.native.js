import React, {useContext} from 'react';
import {View, StyleSheet} from 'react-native';
import {KeyboardContext} from '../../contexts/keyboard/KeyboardContext';
import {UiColors} from '../../contexts/uiColors';
import {VirtualKeyboardElement} from './keyboard';

export const VirtualKeyboard = () => {
  const {isVisible} = useContext(KeyboardContext);
  const stylesThemed = styles(UiColors);

  return (
    isVisible && (
      <View style={stylesThemed.container}>
        <VirtualKeyboardElement />
      </View>
    )
  );
};

// eslint-disable-next-line no-shadow
const styles = UiColors =>
  StyleSheet.create({
    container: {
      position: 'absolute',
      bottom: 0,
      width: '100%',
      minHeight: 50,
      backgroundColor: UiColors.keyboardBg,
    },
  });
