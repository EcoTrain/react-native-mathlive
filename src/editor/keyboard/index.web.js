import React, {useContext} from 'react';
import {View, StyleSheet} from 'react-native';
import {KeyboardContext} from '../../contexts/keyboard/KeyboardContext';
import {UiColors} from '../../contexts/uiColors';
import {VirtualKeyboardElement} from './keyboard';

export const VirtualKeyboard = () => {
  const {isVisible} = useContext(KeyboardContext);
  const stylesThemed = styles(UiColors);

  return (
    <View
      style={[
        stylesThemed.container,
        {
          visibility: isVisible ? 'visible' : 'hidden',
        },
      ]}>
      <VirtualKeyboardElement />
    </View>
  );
};

const styles = UiColors =>
  StyleSheet.create({
    container: {
      position: 'fixed',
      bottom: 0,
      left: 0,
      width: '100%',
      minHeight: 50,
      backgroundColor: UiColors.keyboardBg,
    },
  });
