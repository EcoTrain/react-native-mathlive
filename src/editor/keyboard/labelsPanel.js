import React, {useContext} from 'react';
import {View, Text, StyleSheet, TouchableOpacity} from 'react-native';
import {KeyboardContext} from '../../contexts/keyboard/KeyboardContext';
import {UiColors} from '../../contexts/uiColors';

export const KeyboardTopLabelsPanel = () => {
  const {virtualKeyboards, actualKeyboards} = useContext(KeyboardContext);
  const stylesThemed = styles(UiColors);

  const getKeyboardLabels = () => {
    const labels = [];
    actualKeyboards.map((kbLayoutName, ikb) => {
      const keyboardLayout = virtualKeyboards[kbLayoutName];
      if (keyboardLayout) {
        labels.push(<KeyboardTopLabelItem key={ikb} label={keyboardLayout.label} kbName={kbLayoutName} />);
      }
    });
    return labels;
  };

  return <View style={stylesThemed.labelsPanel}>{getKeyboardLabels()}</View>;
};

const KeyboardTopLabelItem = ({label, kbName}) => {
  const {activeKeyboardName, setActiveKeyboardName} = useContext(KeyboardContext);
  const stylesThemed = styles(UiColors);

  const isActive = activeKeyboardName == kbName;

  return (
    <TouchableOpacity style={stylesThemed.labelPanelItem} onPress={() => setActiveKeyboardName(kbName)}>
      <Text
        style={[
          stylesThemed.labelPanelItemText,
          {
            color: isActive ? UiColors.textSelected : UiColors.textCold,
          },
        ]}>
        {label}
      </Text>
    </TouchableOpacity>
  );
};

// eslint-disable-next-line no-shadow
const styles = UiColors =>
  StyleSheet.create({
    labelsPanel: {
      flex: 1,
      flexDirection: 'row',
    },
    labelPanelItem: {
      padding: 5,
    },
    labelPanelItemText: {
      fontWeight: 'bold',
    },
  });
