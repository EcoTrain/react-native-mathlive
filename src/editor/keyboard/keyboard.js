import React, {useContext} from 'react';
import {StyleSheet, View} from 'react-native';
import {KeyboardContext} from '../../contexts/keyboard/KeyboardContext';
import {UiColors} from '../../contexts/uiColors';
import {KeyboardTopLabelsPanel} from './labelsPanel';
import {KeyboardLayer} from './layer';

export const VirtualKeyboardElement = () => {
  const {virtualKeyboards, virtualKeyboardLayers, activeKeyboardName} = useContext(KeyboardContext),
    stylesThemed = styles(UiColors),
    getActiveKeyboard = () => {
      const activeLayerName = virtualKeyboards[activeKeyboardName]?.layer,
        layer = virtualKeyboardLayers[activeLayerName];
      return <KeyboardLayer layer={layer} />;
    };

  return (
    <View style={stylesThemed.container}>
      <View style={stylesThemed.mainKeyboardPanel}>
        <KeyboardTopLabelsPanel />
        <View>{getActiveKeyboard()}</View>
      </View>
    </View>
  );
};

const styles = UiColors =>
  StyleSheet.create({
    container: {
      alignItems: 'center',
    },
    mainKeyboardPanel: {
      maxWidth: 800,
      width: '100%',
      marginVertical: 10,
      marginHorizontal: 10,
    },
  });
