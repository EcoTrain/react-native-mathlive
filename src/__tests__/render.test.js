import React from 'react';
import {render} from '@testing-library/react-native';
import {Mathfield} from '../App';
import {VirtualKeyboardElement} from '../editor/keyboard/keyboard';

beforeAll(done => {
  done();
});

afterAll(done => {
  done();
});

describe('Mathfield', () => {
  it('Render code correctly', () => {
    const component = render(<Mathfield />);
    expect(component).toMatchSnapshot();
  });
});

describe('VirtualKeyboardElement', () => {
  it('Render code correctly', () => {
    const component = render(<VirtualKeyboardElement />);
    expect(component).toMatchSnapshot();
  });
});
