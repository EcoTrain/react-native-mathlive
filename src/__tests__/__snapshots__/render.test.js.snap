// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mathfield Render code correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "hsl(204, 20%, 98%)",
        "borderColor": "gray",
        "borderRadius": 8,
        "borderWidth": 1,
        "minHeight": 40,
        "opacity": 1,
        "padding": 8,
      }
    }
  >
    <View
      style={
        {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "row",
            "flexWrap": "wrap",
          }
        }
      >
        <View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignContent": "center",
                "backgroundColor": false,
                "height": "100%",
                "justifyContent": "center",
                "opacity": 1,
              }
            }
          >
            <View
              style={
                {
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  [
                    {},
                    false,
                  ]
                }
              >
                3
              </Text>
            </View>
          </View>
        </View>
        <View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignContent": "center",
                "backgroundColor": false,
                "height": "100%",
                "justifyContent": "center",
                "opacity": 1,
              }
            }
          >
            <View
              style={
                {
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  [
                    {},
                    false,
                  ]
                }
              >
                 
              </Text>
            </View>
          </View>
        </View>
        <View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignContent": "center",
                "backgroundColor": false,
                "height": "100%",
                "justifyContent": "center",
                "opacity": 1,
              }
            }
          >
            <View
              style={
                {
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  [
                    {},
                    false,
                  ]
                }
              >
                +
              </Text>
            </View>
          </View>
        </View>
        <View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignContent": "center",
                "backgroundColor": false,
                "height": "100%",
                "justifyContent": "center",
                "opacity": 1,
              }
            }
          >
            <View
              style={
                {
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  [
                    {},
                    false,
                  ]
                }
              >
                 
              </Text>
            </View>
          </View>
        </View>
        <View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignContent": "center",
                "backgroundColor": "hsl(204, 50%, 90%)",
                "height": "100%",
                "justifyContent": "center",
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                [
                  {},
                  false,
                ]
              }
            >
              â–¢
            </Text>
          </View>
        </View>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignSelf": "center",
            "aspectRatio": "1/1",
            "height": "100%",
            "marginLeft": "auto",
            "maxHeight": 40,
            "opacity": 1,
          }
        }
      >
        <Image
          source={
            {
              "process": [Function],
            }
          }
          style={
            {
              "flex": 1,
              "height": "100%",
              "width": "100%",
            }
          }
        />
      </View>
    </View>
  </View>
</View>
`;
